sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c

```diff
@@ -0,0 +1,630 @@
+ */
+mutex_t _list_gnrc_tcp_tcb_lock;
+
+/**
+ * @brief   Establishes a new TCP connection
+ *
+ * @param[in/out] tcb       This connections Transmission control block.
+ * @param[in] target_addr   Target Address to connect to, if this is a active connection.
+ * @param[in] target_port   Target Port to connect to, if this is a active connection.
+ * @param[in] local_addr    Local Address to bind on, if this is a passive connection.
+ * @param[in] local_port    Local Port to bind on, if this is a passive connection.
+ * @param[in] passive       Flag to indicate if this is a active or passive open.
+ *
+ * @return  0 on success.
+ * @return   -EISCONN if transmission control block is already in use.
+ * @return   -ENOMEM if the receive buffer for the tcb could not be allocated.
+ *           Increase "GNRC_TCP_RCV_BUFFERS".
+ * @return   -EADDRINUSE if @p local_port is already used by another connection. Only active mode.
+ * @return   -ETIMEDOUT if the connection could not be opened. Only active mode.
+ * @return   -ECONNREFUSED if the connection was resetted by the peer.
+ */
+static int8_t _gnrc_tcp_open(gnrc_tcp_tcb_t *tcb, const void *target_addr, uint16_t target_port,
+                            const void *local_addr, uint16_t local_port, uint8_t passive)
+{
+    msg_t msg;                           /* Message for incomming Messages */
+    msg_t connection_timeout_msg;        /* Connection Timeout Message */
+    xtimer_t connection_timeout_timer;   /* Connection Timeout Timer */
+    int8_t ret = 0;                      /* Return Value */
+
+    /* Lock the tcb for this function call */
+    mutex_lock(&(tcb->function_lock));
+
+    /* Connection is already connected: Return -EISCONN */
+    if (tcb->state != CLOSED) {
+        mutex_unlock(&(tcb->function_lock));
+        return -EISCONN;
+    }
+
+    /* Setup connection (common parts) */
+    msg_init_queue(tcb->msg_queue, GNRC_TCP_MSG_QUEUE_SIZE);
+    tcb->owner = thread_getpid();
+
+    /* Setup passive connection */
+    if (passive){
+        /* Set Status Flags */
+        tcb->status |= GNRC_TCP_STATUS_PASSIVE;
+        if (local_addr == NULL) {
+            tcb->status |= GNRC_TCP_STATUS_ALLOW_ANY_ADDR;
+        }
+        /* If local address is specified: Copy it into tcb: only connections to this addr are ok */
+        else {
+            switch (tcb->address_family) {
+#ifdef MODULE_GNRC_IPV6
+                case AF_INET6:
+                    memcpy(tcb->local_addr, local_addr, sizeof(ipv6_addr_t));
+                    break;
+#endif
+            }
+        }
+        /* Assign Port to listen on, to tcb */
+        *(tcb->local_port) = local_port;
+    }
+    /* Setup active connection */
+    else{
+        /* Copy Target Address and Port into tcb structure */
+        switch (tcb->address_family) {
+ #ifdef MODULE_GNRC_IPV6
+            case AF_INET6:
+                memcpy(tcb->peer_addr, target_addr, sizeof(ipv6_addr_t));
```

Comment:
True. Althrough it is stated in the functions documantation that target_addr must not be NULL.